firewood/src/storage/persisted/scheduler.rs

```diff
@@ -0,0 +1,138 @@
+use super::{PersistedStore, StoreWrite};
+use std::sync::Arc;
+use thiserror::Error;
+use tokio::sync::{mpsc, Notify};
+
+#[derive(Debug, Error)]
+pub enum StoreError<T> {
+    #[error("error sending data: `{0}`")]
+    Send(#[from] tokio::sync::mpsc::error::SendError<T>),
+}
+
+struct DiskRequester {
+    sender: mpsc::Sender<(Vec<StoreWrite>, Arc<Notify>)>,
+}
+
+impl DiskRequester {
+    const fn new(write_sender: mpsc::Sender<(Vec<StoreWrite>, Arc<Notify>)>) -> Self {
+        Self {
+            sender: write_sender,
+        }
+    }
+
+    pub async fn send_writes(&self, writes: Vec<StoreWrite>) {
+        let notifier = Arc::new(Notify::new());
+
+        self.sender
+            .send((writes, notifier.clone()))
+            .await
+            .map_err(StoreError::Send)
+            .ok();
+
+        // wait on the notification
+        notifier.notified().await
+    }
+}
+
+struct DiskScheduler {
+    receiver: mpsc::Receiver<(Vec<StoreWrite>, Arc<Notify>)>,
+}
+
+impl DiskScheduler {
+    const fn new(write_receiver: mpsc::Receiver<(Vec<StoreWrite>, Arc<Notify>)>) -> Self {
+        Self {
+            receiver: write_receiver,
+        }
+    }
+
+    #[tokio::main(flavor = "current_thread")]
+    // The caller (e.g. Coordinator) should spawn the writer thread.
+    pub async fn run(self, store: &mut PersistedStore) {
+        let mut receiver = self.receiver;
+        let local_pool = tokio::task::LocalSet::new();
+        local_pool
+            // everything needs to be moved into this future in order to be properly dropped
+            .run_until(async move {
+                loop {
+                    while let Some(element) = receiver.recv().await {
+                        // Write to the store
+                        let (writes, notifier) = element;
+                        store.write_all(&writes).await.unwrap_or_else(|err| {
+                            eprintln!("Error writing to store: {:?}", err);
+                        });
+
+                        // Ack the writes are complete.
+                        notifier.notify_one();
+                    }
+                }
+            })
+            .await;
+
+        // when finished process all requests, wait for any pending-futures to complete
+        local_pool.await;
```

Comment:
How do we get here? `run_until` will block until the future completes, which is in a `loop` with no `break` statement.